packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -212,7 +212,7 @@ function proxyESM(id: string, mod: any) {
   return new Proxy(mod, {
     get(mod, prop) {
       if (prop === 'default') return defaultExport
-      return mod[prop]
+      return mod[prop] ?? defaultExport?.[prop]
```

Comment:
This approach is fragile. If a named export equals undefined or null, the default export is assumed to be an object, but we can't guarantee that.

#5197 was using this logic in `proxyESM`...
```ts
if (prop in mod) {
  return mod[prop]
}
// commonjs interop: module whose exports are not statically analyzable
if (isObject(mod.default) && prop in mod.default) {
  return mod.default[prop]
}
// throw an error like ESM import does
throw new SyntaxError(
  `The requested module '${id}' does not provide an export named '${prop.toString()}'`
)
```
...until @matthewp reverted it without explanation.

Reply:
Thanks for digging into this, @natemoo-re @matthewp could you check why this was removed? Maybe as part of the trials to make CI pass? Let's add this back before releasing 2.7

Reply:
I'm not sure why this was removed, but this approach LGTM.

Reply:
The throw was a problem there. Since `mod` is an object, any code that imports the namespace can inspect it for keys just like you can do with any object. The throw assumed that the `get` always comes from a static import statement but that's not necessarily the case:

```js
// This is valid code
import * as mod from './something.js';

if(mod.someKeyThatDoesntExist) {
  // Oops this threw
}
```

Also the second condition looks wrong to me too. Given this example:

__mod.js__

```js
export default { color: 'green' };
```

__main.js__

```js
import * as mod from './mod.js';

console.log(mod.color);
```

Should log `undefined` but will log 'green' instead. If commonjs interop is needed it needs to verify that the module is commonjs first.